examples/matching-engine/tests/transaction.rs

```diff
@@ -0,0 +1,263 @@
+        orders_bids.extend(order_messages);
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_bids);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+    matching_chain.handle_received_messages().await;
+
+    // Checking the values for token_a
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::from_tokens(1))),
+        (owner_b, None),
+    ] {
+        let value = FungibleTokenAbi::query_account(token_id_a, &fungible_chain_a, owner).await;
+        assert_eq!(value, amount);
+    }
+    for (owner, amount) in [(admin_account, None), (owner_a, None), (owner_b, None)] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &fungible_chain_b, owner).await,
+            amount
+        );
+    }
+    for (owner, amount) in [
+        (admin_account, None),
+        (owner_a, Some(Amount::ZERO)),
+        (owner_b, None),
+    ] {
+        assert_eq!(
+            FungibleTokenAbi::query_account(token_id_a, &matching_chain, owner).await,
+            amount
+        );
+    }
+
+    let mut orders_asks = Vec::new();
+    for price in [4, 2] {
+        let price = Price { price };
+        let order = Order::Insert {
+            owner: owner_b,
+            amount: Amount::from_tokens(4),
+            nature: OrderNature::Ask,
+            price,
+        };
+        let operation = Operation::ExecuteOrder { order };
+        let order_messages = fungible_chain_b
+            .add_block(|block| {
+                block.with_operation(matching_id, operation);
+            })
+            .await;
+
+        assert_eq!(order_messages.len(), 3);
+        orders_asks.extend(order_messages);
+    }
+
+    matching_chain
+        .add_block(|block| {
+            block.with_incoming_messages(orders_asks);
+        })
+        .await;
+
+    fungible_chain_a.handle_received_messages().await;
+    fungible_chain_b.handle_received_messages().await;
+    matching_chain.handle_received_messages().await;
+
+    // Now querying the matching
+    let order_ids_a = get_orders(matching_id, &matching_chain, owner_a)
+        .await
+        .expect("order_ids_a");
+    assert_eq!(order_ids_a.len(), 1);
+    let order_ids_b = get_orders(matching_id, &matching_chain, owner_b)
+        .await
+        .expect("order_ids_b");
+    assert_eq!(order_ids_b.len(), 2);
+
+    // Cancelling the remaining orders
+    let mut orders_cancels = Vec::new();
+    for (owner, order_ids) in [(owner_a, order_ids_a), (owner_b, order_ids_b)] {
+        for order_id in order_ids {
+            let order = Order::Cancel { owner, order_id };
+            let operation = Operation::ExecuteOrder { order };
+            let order_messages = fungible_chain_a
+                .add_block(|block| {
+                    block.with_operation(matching_id, operation);
```

Comment:
Why is A's chain allowed to cancel B's order?

Reply:
It is because for a matching engine only one chain really matters and so the cancellation order can be sent to any chain, it is then transmitted to the matching engine chain.

Reply:
We want user authentication for all orders. Let's fix this in a next PR (should be easy)